### PR TITLE
add initial parallel test runner

### PR DESCRIPTION
**Pull request by qmx**
this introduces a parallel runner for our test harness

### CODE REVIEW

**Comment by qmx** [comment]
marked as WIP cause I still need to fix the verbose output mode

File: src/bin/tester.rs [hunk 1]
```diff
@@ -136,6 +139,87 @@ fn ambiguity_check(handle: ModuleContentsHandle) -> Result<(), MoreThanOne> {
     Ok(())
 }
 
+#[derive(Debug)]
+struct Counters {
```
Comment (cramertj) [inline src/bin/tester.rs:179]: Nit: `derive(Default)` as well to get easier initialize-with-zeros below

File: src/bin/tester.rs [hunk 2]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
         Command::Dir { verbose, dir } => {
-            // Counters for reporting overall stats at the end.
-            let mut total_count = 0;
-            let mut unambiguous_count = 0;
-            let mut ambiguous_count = 0;
-            let mut too_short_count = 0;
-            let mut no_parse_count = 0;
+            // HACK(eddyb) avoid parsing some files that hit
+            // `lykenware/gll` worst-cases (many GBs of RAM usage)
+            // FIXME(eddyb) fix the problems (e.g. implement GC).
```
Comment (cramertj) [inline src/bin/tester.rs]: oh dear lol

File: src/bin/tester.rs [hunk 3]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
-                            ambiguity_result = ambiguity_check(handle);
-                            if ambiguity_result.is_ok() {
-                                ('.', &mut unambiguous_count)
-                            } else {
-                                ('-', &mut ambiguous_count)
+            let mut counters: Counters = files
+                .par_bridge()
+                .map(|f| process(f, verbose))
+                .fold(
+                    || Counters {
```
Comment (cramertj) [inline src/bin/tester.rs]: replace with `Counters::default()`

File: src/bin/tester.rs [hunk 4]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
+                            }
+                            ParseResult::Error => {
+                                acc.no_parse_count += 1;
+                            }
+                        };
+                        acc
+                    },
+                )
+                .reduce(
+                    || Counters {
```
Comment (cramertj) [inline src/bin/tester.rs]: `Counters::default()` again

File: src/bin/tester.rs [hunk 5]
```diff
@@ -136,6 +139,87 @@ fn ambiguity_check(handle: ModuleContentsHandle) -> Result<(), MoreThanOne> {
...
+#[derive(Debug)]
+struct Counters {
+    total_count: u16,
+    unambiguous_count: u16,
+    ambiguous_count: u16,
+    too_short_count: u16,
+    no_parse_count: u16,
+}
+
+impl Add for Counters {
```
Comment (Centril) [inline src/bin/tester.rs]: Use https://jeltef.github.io/derive_more/derive_more/add.html ?

File: src/bin/tester.rs [hunk 6]
```diff
@@ -136,6 +139,87 @@ fn ambiguity_check(handle: ModuleContentsHandle) -> Result<(), MoreThanOne> {
...
+#[derive(Debug)]
+enum ParseResult {
+    Unambiguous,
+    Ambiguous,
+    Partial,
+    Error,
+}
+
+impl ParseResult {
+    fn compact_display(&self) -> String {
```
Comment (Centril) [inline src/bin/tester.rs]: Return a `&'static str` and `.to_string()` where needed instead (probably not needed to do `.to_string()` actually)?

File: src/bin/tester.rs [hunk 7]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
-            // Counters for reporting overall stats at the end.
-            let mut total_count = 0;
-            let mut unambiguous_count = 0;
-            let mut ambiguous_count = 0;
-            let mut too_short_count = 0;
-            let mut no_parse_count = 0;
+            // HACK(eddyb) avoid parsing some files that hit
+            // `lykenware/gll` worst-cases (many GBs of RAM usage)
+            // FIXME(eddyb) fix the problems (e.g. implement GC).
+            const BLACKLIST: &[&str] = &[
```
Comment (Centril) [inline src/bin/tester.rs]: Extract out to top level for better visibility (or a separate config file...?)?

File: src/bin/tester.rs [hunk 8]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
             // Find all the `.rs` files inside the desired directory.
             let files = WalkDir::new(dir)
                 .contents_first(true)
                 .into_iter()
                 .map(|entry| entry.unwrap())
-                .filter(|entry| entry.path().extension().map_or(false, |ext| ext == "rs"));
-
-            let mut stdout = io::stdout();
+                .filter(|entry| entry.path().extension().map_or(false, |ext| ext == "rs"))
+                .filter(|entry| !BLACKLIST.iter().any(|&b| entry.path().ends_with(b)));
```
Comment (Centril) [inline src/bin/tester.rs]: I'd extract out to a function `is_blacklisted` so we can iterate on the blacklisting mechanism there.

File: src/bin/tester.rs [hunk 9]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
+                        ambiguous_count: 0,
+                        unambiguous_count: 0,
+                        too_short_count: 0,
+                        no_parse_count: 0,
+                    },
+                    |mut acc, x| {
+                        acc.total_count += 1;
+                        match x {
+                            ParseResult::Ambiguous => {
+                                acc.ambiguous_count += 1;
```
Comment (Centril) [inline src/bin/tester.rs:307]: Not sure if better, but you could replace this with `=> &mut acc.ambiguous_count,` (for each arm + field) and then increment the returned field.

File: src/bin/tester.rs [hunk 10]
```diff
@@ -163,92 +247,85 @@ fn main() {
...
+                        total_count: 0,
+                        ambiguous_count: 0,
+                        unambiguous_count: 0,
+                        too_short_count: 0,
+                        no_parse_count: 0,
+                    },
+                    |a, b| a + b,
+                );
 
             // We're done, time to print out stats!
```
Comment (Centril) [inline src/bin/tester.rs:324]: Extract out to a function `print_statistics(counters);`?

**Review by Centril — COMMENTED** [review]
Assorted suggestions :)

File: src/bin/tester.rs (hunk 5 above)
Comment (qmx) [reply src/bin/tester.rs]: Wow, I didn't know this one! super cool!

File: src/bin/tester.rs (hunk 7 above)
Comment (qmx) [reply src/bin/tester.rs]: yea, this one is going to a config file

File: src/bin/tester.rs (hunk 9 above)
Comment (qmx) [reply src/bin/tester.rs:307]: the original code was doing this and it took me a while to figure out what it was doing, hence I've opted for the more explicit version

File: src/bin/tester.rs [hunk 11]
```diff
@@ -6,18 +6,51 @@ extern crate structopt;
...
+struct Config {
+    blacklist: Blacklist,
+}
+
+impl Config {
+    fn load() -> Result<Config, failure::Error> {
+        let config = match File::open("wg-grammar.toml") {
+            Ok(mut file) => {
+                let mut toml = String::new();
+                file.read_to_string(&mut toml)?;
```
Comment (Centril) [inline src/bin/tester.rs]: Use https://doc.rust-lang.org/nightly/std/fs/fn.read_to_string.html?

**Review by Centril — APPROVED** [review]

File: src/bin/tester.rs (hunk 7 above)
Comment (eddyb) [reply src/bin/tester.rs]: The HACK/FIXME comments were lost when moving to a separate file.

Comment (qmx) [reply src/bin/tester.rs]: this was intentional from my part, now that the blacklist lives externally on the config file, and we have an issue for tracking the memory leak/lack of GC - we can bring that back if you want.

File: src/bin/tester.rs [hunk 12]
```diff
@@ -136,7 +175,88 @@ fn ambiguity_check(handle: ModuleContentsHandle) -> Result<(), MoreThanOne> {
...
+    }
+}
+
+fn process(file: walkdir::DirEntry, verbose: bool) -> ParseResult {
+    let mut stdout = io::stdout();
+    let path = file.into_path();
+
+    let out = parse_file_with(&path, |result| {
+        let mut ambiguity_result = Ok(());
+        let start = Instant::now();
```
Comment (eddyb) [inline src/bin/tester.rs:212]: This is after the parse, it only measures `ambiguity_check` (which is a graph traversal, pretty insignificant)

**Pull request merged**